Gems/Archive/Code/Source/Clients/ArchiveReader.cpp

```diff
@@ -0,0 +1,885 @@
+            const size_t availableBytesInCompressedBlock = AZStd::min(compressedBlockRemainingSpan.size(),
+                ArchiveBlockSizeForCompression);
+            AZStd::span<const AZStd::byte> compressedDataForBlock = compressedBlockRemainingSpan
+                .first(availableBytesInCompressedBlock)
+                .first(blockCompressedSize);
+            // Slide the compressed block remaining span by 2 MiB
+            compressedBlockRemainingSpan.subspan(availableBytesInCompressedBlock);
+
+            // Get the block span for storing the decompressed block
+            // As the uncompressed size is 2 MiB for all blocks except the last
+            // the entire contiguous file sequence will be available in the decompressedResultSpan after the loop
+            const size_t remainingBytesInBlockSpan = AZStd::min(decompressionRemainingSpan.size(),
+                ArchiveBlockSizeForCompression);
+            AZStd::span<AZStd::byte> decompressionBlockSpan = decompressionRemainingSpan
+                .first(remainingBytesInBlockSpan);
+            // Slide the remaining decompressed span by 2 MiB as well
+            decompressionRemainingSpan.subspan(remainingBytesInBlockSpan);
+
+            const AZ::u32 decompressTaskCount = AZStd::min(static_cast<AZ::u32>(blockRange.second - blockRange.first),
+                maxDecompressTasks);
+
+            // Task graph event used to block decompressing blocks in parallel
+            auto taskDecompressGraphEvent = AZStd::make_unique<AZ::TaskGraphEvent>("Content File Decompress Sync");
+            AZ::TaskGraph taskGraph{ "Archive Decompress Tasks" };
+            AZ::TaskDescriptor decompressTaskDescriptor{ "Decompress Block", "Archive Content File Decompression" };
+            for (AZ::u32 decompressTaskSlot = 0; decompressTaskSlot < decompressTaskCount; ++decompressTaskSlot)
+            {
+                //! Decompress Task to execute in task executor
+                auto decompressTask = [
+                    decompressionInterface, &decompressionOptions, decompressionBlockSpan,
+                        compressedDataForBlock,
+                        &decompressedBlockResult = decompressedBlockResults[decompressTaskSlot]]()
+                {
+                    // Decompressed the compressed block
+                    decompressedBlockResult = decompressionInterface->DecompressBlock(
+                        decompressionBlockSpan, compressedDataForBlock, decompressionOptions);
+                };
+
+                    taskGraph.AddTask(decompressTaskDescriptor, AZStd::move(decompressTask));
+                    // Increment to the block index
+                    ++blockRange.first;
+            }
+
+            taskGraph.SubmitOnExecutor(m_taskExecutor, taskDecompressGraphEvent.get());
+            // Sync on the task completion
+            taskDecompressGraphEvent->Wait();
```

Comment:
Is the sync here where we wait on all tasks in the task graph to complete preferred to using a work queue to avoid the locks on the queue and be faster overall?  I'm not recommending changing the approach in this PR, just want to understand the tradeoffs.

Reply:
Well in this case the number of decompression task is limited by the lower between the number of blocks remaining to decompress into a 2-MIB chunk of the file and the core count of the system.

Task slots are preallocated up front and only run on a block of a compressed file that are segmented to not overlap.

The sync is preferable here so that that the next chunk of blocks being decompressed don't have to search for an available task slot.
Otherwise a synchronization primitive would have to be used to keep track of the active set of decompression jobs that are being run.